Enable getting of Access Token in iOS and Android.

seaders opened the pull request.

seaders commented:
I signed it!

Corporation/Company name: SixMinute.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSAndroidSetupUI.cs:
```diff
@@ -21,6 +21,9 @@ namespace GooglePlayGames
 
     public class GPGSAndroidSetupUI : EditorWindow
     {
+		private const string GameInfoPath = "Assets/GooglePlayGames/GameInfo.cs";
```
nit: indentation should be consistent.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSAndroidSetupUI.cs:
```diff
@@ -28,12 +31,13 @@ public static void MenuItemFileGPGSAndroidSetup()
         public void OnEnable()
         {
             mAppId = GPGSProjectSettings.Instance.Get("proj.AppId");
+            mClientId = GPGSProjectSettings.Instance.Get(GPGSUtil.ANDROIDCLIENTIDKEY);
```
This is the client id of the server, correct?

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSAndroidSetupUI.cs:
```diff
@@ -86,14 +100,26 @@ public void DoSetup()
-        public static bool PerformSetup(string appId, string nearbySvcId)
-        {
-            // check for valid app id
-            if (!GPGSUtil.LooksLikeValidAppId(appId))
```
Can this be optional?  So if the game does not need the token, this does not need to be configured? (And it will be backwards compatible).

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSAndroidSetupUI.cs:
```diff
@@ -104,7 +130,9 @@ public static bool PerformSetup(string appId, string nearbySvcId)
             }
 
             GPGSProjectSettings.Instance.Set("proj.AppId", appId);
+			GPGSProjectSettings.Instance.Set(GPGSUtil.ANDROIDCLIENTIDKEY, clientId);
```
nit: indentation should be consistent.

claywilkinson replied:
If it is easy to fix the indentation, please do, but if it is a pain, don't worry about it.

seaders commented:
Yeah, sorry about that.  I tried my best but MonoDevelop kept on messing things up.  You want me to correct, or can you do it easy enough?

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSAndroidSetupUI.cs:
```diff
@@ -127,5 +157,21 @@ public static bool PerformSetup(string appId, string nearbySvcId)
 
             return true;
         }
+		
```
Can this be merged into GPGSUtil.UpdateGameInfo so it is being updated only from one method?

seaders replied:
It definitely could.  I wasn't sure exactly what strategy you might have wanted to go with that route, whether to keep them separate (as the Android process was quite dissimilar to this one before these changes).

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSStrings.cs:
```diff
@@ -45,6 +45,11 @@ public class Setup
             public const string AppIdError = "The App Id does not appear to be valid. " +
                                              "It must consist solely of digits, usually 10 or more.";
+			
+			public const string ClientIdTitle = "Oauth Client ID";
```
nit: indentation should be consistent.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSStrings.cs:
```diff
@@ -95,6 +95,10 @@ public class AndroidSetup
             public const string Blurb = "To configure Google Play Games in this project,\n" +
                                         "please enter the information below and click on the Setup button.";
 
+            public const string ClientIdBlurb = "Enter your oauth2 client ID below. To obtain this\n" +
```
duplicate string from line 70?

seaders replied:
They are different things, 70-72 is mentioning iOS, 98-100 are Android.  I'm more than happy to change to something else, but I think they may need slightly different descriptions for what they're logically referring to.

claywilkinson replied:
ahh sorry about that - read it too fast.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/GPGSUtil.cs:
```diff
@@ -27,10 +27,14 @@ public static class GPGSUtil
-        public const string IOSCLIENTIDKEY = "ios.ClientId";
-        public const string IOSBUNDLEIDKEY = "ios.BundleId";
+
+		public const string ANDROIDCLIENTIDPLACEHOLDER = "__ANDROID_CLIENTID__";
```
nit: indentation should be consistent.

seaders commented:
Re: the clientId (I can't find where your comment is now), it is optional.  It only processes it if it's not null, otherwise it reverts back to how it used to work.  And the clientId is the service id, but you get it basically just the same as you get the one for iOS.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Editor/template-AndroidManifest.txt:
```diff
@@ -8,6 +8,9 @@
     
     <uses-sdk android:minSdkVersion="8" android:targetSdkVersion="16" /> 
  
+    <uses-permission android:name="android.permission.GET_ACCOUNTS"/>
```
can we make these optionally added?  for example if the client id is not entered (or adda checkbox to the setup dialog?) then add these.  It seems heavy handed to make every game add these permissions.

seaders replied:
Yeah, this was one of those things that I couldn't see exactly how to get the template to have optional things like this in.  If possible, do you think you could work the optionality in?

claywilkinson replied:
sure.  I think the easiest way is to have a static string with the permission elements and then a placeholder in the template which is either replaced with the static string or an empty string.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/ISocialPlatform/PlayGamesLocalUser.cs:
```diff
@@ -132,6 +132,20 @@ public new string id
+        {
+            get
+            {
+                return authenticated ? mPlatform.GetToken() : string.Empty;
```
should return null?  this way it is consistent with PlayGamesPlatform.GetToken()

seaders replied:
Maybe, I was working off the consistencies in this file, id and userName et al.

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Platforms/Native/NativeClient.cs:
```diff
@@ -285,6 +285,71 @@ private static void LaunchBridgeIntent(IntPtr bridgedIntent)
             }
         }
+        
+        private AndroidJavaObject GetApiClient(GameServices services) {
```
Can you replace this with
`using C = GooglePlayGames.Native.Cwrapper.InternalHooks;`

`return JavaUtils.JavaObjectFromPointer(C.InternalHooks_GetApiClient(services.AsHandle()));`

seaders replied:
Replace the whole function?  As in to,


        private AndroidJavaObject GetApiClient(GameServices services) {
            using (var C = GooglePlayGames.Native.Cwrapper.InternalHooks)
            {
                return JavaUtils.JavaObjectFromPointer(C.InternalHooks_GetApiClient(services.AsHandle()));
            }
        }

?

claywilkinson replied:
yep, actually at the top of the file with the using statements.  You can look at GooglePlayGames/Platforms/Native/AndroidAppStateClient.cs for an example

claywilkinson commented on source/PluginDev/Assets/GooglePlayGames/Platforms/Native/NativeClient.cs:
```diff
@@ -285,6 +285,71 @@ private static void LaunchBridgeIntent(IntPtr bridgedIntent)
+            Debug.Log("Before RetrieveUserEmail");
+            string email = RetrieveUserEmail() ?? "NULL";
+            Debug.Log("After RetrieveUserEmail email: " + email);
+            string scope = "audience:server:client_id:" + "101626759741-kc5sdafasdfsdf9j1aek9bfgfou3oom.apps.googleusercontent.com";//CLIENT_ID;
```
This should be the GameInfo.AndroidClientId, not the hardcoded one?

seaders replied:
100%, will be in the next commit.

claywilkinson commented:
Thanks for the PR!! I made a few comments - nothing major - let me know when I should merge it.

claywilkinson commented:
I confirmed that the CLA for Sixminute was signed.

seaders commented:
Gimme about an hour, and I'll get all your requests in.

seaders commented:
You need a brand new pull request, or can https://github.com/SixMinute/play-games-plugin-for-unity/commit/3fe84d7735b3ed23acd0e5fd7647e6a9ceb26740 be brought in to this?

gguuss commented on source/PluginDev/Assets/GooglePlayGames/Platforms/Native/NativeClient.cs:
```diff
@@ -285,6 +286,56 @@ private static void LaunchBridgeIntent(IntPtr bridgedIntent)
+            Debug.Log("Before RetrieveUserEmail");
+            string email = RetrieveUserEmail() ?? "NULL";
+            Debug.Log("After RetrieveUserEmail email: " + email);
+            string scope = "audience:server:client_id:" + GameInfo.AndroidClientId;
```
Note that this will not return an access token, but instead an OpenID connect ID token - getting the ID token is actually what you should be doing. That said, to get the access token, specify the scope you're trying to retrieve, e.g. `https://www.googleapis.com/auth/plus.me` and exclude the server audience in the scope:

      /// <summary>Gets the access token currently assocuiated with the Unity activity.</summary>
      /// <returns>The OAuth 2.0 access token.</returns>
      public string GetAccessToken()
      {
          string token = null;
          string email = GetEmail() ?? "NULL";
          string scope = "oauth2:https://www.googleapis.com/auth/plus.me";

          using (AndroidJavaClass unityActivity = new AndroidJavaClass("com.unity3d.player.UnityPlayer"),
                 googleAuthUtil = new AndroidJavaClass("com.google.android.gms.auth.GoogleAuthUtil"))
          {
              using(AndroidJavaObject currentActivity =
                  unityActivity.GetStatic<AndroidJavaObject>("currentActivity"))
              {
                  token = googleAuthUtil.CallStatic<string>("getToken", currentActivity, email, scope);
              }
          }

          return token;
      }

seaders replied:
If possible, I'd love for you guys to own that side of things and offer those options in a way you want to users.  I put this in based on @claywilkinson's feedback in the thread stating this looked great and asked for a pull request.

Potentially pull to a separate branch, then make the changes you want there.  I, like so many others in the thread, needed to get access to a access token that could be used to authenticate server-side and this 100% serves my purposes, and I'm assuming many of the other guys too.  This was more to "plug the gap" of not having this functionality at all, if you now want to take it and make it much, much better, I'd be delighted with that.  But I don't necessarily have the bandwidth to make it 100% bulletproof and exactly how Google want it to be.  I think that probably should be Google's responsibility.

gguuss replied:
Sounds good, we'll be happy to make the changes after the merge. Thanks again for your contributions :)

claywilkinson replied:
This is a good point. @seaders  - can you rename your GetToken to GetIDToken() - since these are very different tokens, and then we can add the GetAccessToken()?

seaders commented:
Done in https://github.com/SixMinute/play-games-plugin-for-unity/commit/5e5601ac65a1867dfac2f26267726b8eb4d9ad68.

seaders commented:
After building, I hadn't actually caught them all, but have done now in https://github.com/SixMinute/play-games-plugin-for-unity/commit/a04404c3189bceb0d8d8692ae377edef31b2d35d.

claywilkinson commented:
Thanks again for the updates.  Merging it now.  I feel strongly about the optional permission part, so once I add that, I'll rebuild the unitypackage.

seaders commented:
Understandable.  May think about it being a straight up check box, rather than checking the client id?  Reason I'm saying that is @gguuss's real `GetAccessToken`, that doesn't require the client id, so that shouldn't be required to enable it.

Also, I've just checked the change for GetApiClient, but get,

    07-30 21:45:52.266  24174-24193/? I/Unity﹕ AndroidJavaException: java.lang.IllegalStateException: GoogleApiClient is not configured to use the Plus.API Api. Pass this into GoogleApiClient.Builder#addApi() to use this feature.
            at UnityEngine.AndroidJNISafe.CheckException () [0x00000] in <filename unknown>:0
            at UnityEngine.AndroidJNISafe.CallStringMethod (IntPtr obj, IntPtr methodID, UnityEngine.jvalue[] args) [0x00000] in <filename unknown>:0
            at UnityEngine.AndroidJavaObject._Call[String] (System.String methodName, System.Object[] args) [0x00000] in <filename unknown>:0
            at UnityEngine.AndroidJavaObject.Call[String] (System.String methodName, System.Object[] args) [0x00000] in <filename unknown>:0
            at GooglePlayGames.Native.NativeClient.GetEmail () [0x00000] in <filename unknown>:0
            at GooglePlayGames.Native.NativeClient.GetAccessToken () [0x00000] in <filename unknown>:0
            at GooglePlayGames.PlayGamesPlatform.GetAccessToken () [0x00000] in <filename unknown>:0
            at MainScript.setToken () [0x00000] in <filename unknown>:0

So I think we still need to add the API.

Running a few tests now, will update after confirmation.

seaders commented:
Ok, just ran those.  You *don't* need the permissions for @gguuss's `GetAccessToken`, but you still need the old `GetApiClient` version which adds the `API` and `scope` of `API` and `SCOPE_PLUS_LOGIN` respectively.

I have added in the AccessToken function (including iOS) here, https://github.com/SixMinute/play-games-plugin-for-unity/commit/90111007a9184f2bd9e2f75984a62d1a62b08c4c and fixed a bug in the Android Setup screen.

Sorry for the multiple commits, but my work environment's a little messier than I like at the moment, so I can't run perfectly rigorous checks as I normally would.

claywilkinson commented:
No problem - thanks for testing :)
I already merged this Pull Request, so you'll need to create a new one.

seaders commented:
Doneso!  (ignore the wrong one, that was a mess-up on email addresses)
